warp/src/tesseract/wasm.rs

```diff
@@ -0,0 +1,105 @@
+use crate::tesseract::{self, TesseractEvent};
+use futures::{stream::BoxStream, StreamExt};
+use wasm_bindgen::prelude::*;
+
+#[wasm_bindgen]
+#[derive(Default)]
+pub struct Tesseract {
+    inner: tesseract::Tesseract,
+}
+
+#[wasm_bindgen]
+impl Tesseract {
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Tesseract {
+        Tesseract::default()
+    }
+
+    pub fn set_autosave(&self) {
+        self.inner.set_autosave();
+    }
+
+    pub fn autosave_enabled(&self) -> bool {
+        self.inner.autosave_enabled()
+    }
+
+    pub fn disable_key_check(&self) {
+        self.inner.disable_key_check();
+    }
+
+    pub fn enable_key_check(&self) {
+        self.inner.enable_key_check();
+    }
+
+    pub fn is_key_check_enabled(&self) -> bool {
+        self.inner.is_key_check_enabled()
+    }
+
+    pub fn set(&self, key: &str, value: &str) -> Result<(), JsError> {
+        self.inner.set(key, value).map_err(|e| e.into())
+    }
+
+    pub fn exist(&self, key: &str) -> bool {
+        self.inner.exist(key)
+    }
+
+    pub fn retrieve(&self, key: &str) -> Result<String, JsError> {
+        self.inner.retrieve(key).map_err(|e| e.into())
+    }
+
+    pub fn update_unlock(
+        &self,
+        old_passphrase: &[u8],
+        new_passphrase: &[u8],
+    ) -> Result<(), JsError> {
+        self.inner
+            .update_unlock(old_passphrase, new_passphrase)
+            .map_err(|e| e.into())
+    }
+
+    pub fn delete(&self, key: &str) -> Result<(), JsError> {
+        self.inner.delete(key).map_err(|e| e.into())
+    }
+
+    pub fn clear(&self) {
+        self.inner.clear();
+    }
+
+    pub fn is_unlock(&self) -> bool {
+        self.inner.is_unlock()
+    }
+
+    pub fn unlock(&self, passphrase: &[u8]) -> Result<(), JsError> {
+        self.inner.unlock(passphrase).map_err(|e| e.into())
+    }
+
+    pub fn lock(&self) {
+        self.inner.lock();
+    }
+
+    pub fn save(&self) -> Result<(), JsError> {
+        self.inner.save().map_err(|e| e.into())
+    }
+
+    pub fn subscribe(&self) -> Subscription {
+        Subscription {
+            inner: self.inner.subscribe(),
+        }
+    }
+
+    pub fn load_from_storage(&self) -> Result<(), JsError> {
+        self.inner.load_from_storage().map_err(|e| e.into())
+    }
+}
+
+#[wasm_bindgen]
+pub struct Subscription {
+    inner: BoxStream<'static, TesseractEvent>,
+}
+
+#[wasm_bindgen]
+impl Subscription {
+    pub async fn next(&mut self) -> Option<TesseractEvent> {
+        self.inner.next().await
```

Comment:
I think the problem here is that we would be continuously polling a stream and if that stream returns `None` (or `Poll::Ready(None)`), we should stop polling otherwise it would in many cases be considered as a UB (eg if tesseract completely drops (as in no other references), the sender used would also drop, causing the receiver to close too). If we are to go this route, we might want to see if there is a way for JS to break that loop so it no longer polls if it returns `None` from rust. Cannot recall the equiv translation that wasm-bindgen export would unless it just returns as a `null`, in which case we could probably try that. Ideally though, it would be best to try to use an asynciterator or maybe tap into the event listener from js so the events are subscribe there.

Reply:
I managed to restructure it to satisfy what's expected by js for an asynciterator.